File: src/css/values/resolution.zig
pub const css = @import("../css_parser.zig");
const Result = css.Result;
const Printer = css.Printer;
const PrintErr = css.PrintErr;
const CSSNumber = css.css_values.number.CSSNumber;

/// A CSS `<resolution>` value.
pub const Resolution = union(enum) {
    /// A resolution in dots per inch.
    dpi: CSSNumber,
    /// A resolution in dots per centimeter.
    dpcm: CSSNumber,
    /// A resolution in dots per px.
    dppx: CSSNumber,

    // ~toCssImpl
    const This = @This();

    pub fn hash(this: *const @This(), hasher: *std.hash.Wyhash) void {
        return css.implementHash(@This(), this, hasher);
    }

    pub fn eql(this: *const Resolution, other: *const Resolution) bool {
        return css.implementEql(@This(), this, other);
    }

    pub fn parse(input: *css.Parser) Result(Resolution) {
        // TODO: calc?
        const location = input.currentSourceLocation();
        const tok = switch (input.next()) {
            .result => |vv| vv,
            .err => |e| return .{ .err = e },
        };
        if (tok.* == .dimension) {
            const value = tok.dimension.num.value;
            const unit = tok.dimension.unit;
            // css.todo_stuff.match_ignore_ascii_case
            if (bun.strings.eqlCaseInsensitiveASCIIICheckLength(unit, "dpi")) return .{ .result = .{ .dpi = value } };
            if (bun.strings.eqlCaseInsensitiveASCIIICheckLength(unit, "dpcm")) return .{ .result = .{ .dpcm = value } };
            if (bun.strings.eqlCaseInsensitiveASCIIICheckLength(unit, "dppx") or bun.strings.eqlCaseInsensitiveASCIIICheckLength(unit, "x")) return .{ .result = .{ .dppx = value } };
            return .{ .err = location.newUnexpectedTokenError(.{ .ident = unit }) };
        }
        return .{ .err = location.newUnexpectedTokenError(tok.*) };
    }

    pub fn tryFromToken(token: *const css.Token) css.Maybe(Resolution, void) {
        switch (token.*) {
            .dimension => |dim| {
                const value = dim.num.value;
                const unit = dim.unit;
                // todo_stuff.match_ignore_ascii_case
                if (bun.strings.eqlCaseInsensitiveASCIIICheckLength(unit, "dpi")) {
                    return .{ .result = .{ .dpi = value } };
                } else if (bun.strings.eqlCaseInsensitiveASCIIICheckLength(unit, "dpcm")) {
                    return .{ .result = .{ .dpcm = value } };
                } else if (bun.strings.eqlCaseInsensitiveASCIIICheckLength(unit, "dppx") or
                    bun.strings.eqlCaseInsensitiveASCIIICheckLength(unit, "x"))
                {
                    return .{ .result = .{ .dppx = value } };
                } else {
                    return .{ .err = {} };
                }
            },
            else => return .{ .err = {} },
        }
    }

    pub fn toCss(this: *const This, dest: *Printer) PrintErr!void {
        const value, const unit = switch (this.*) {
            .dpi => |dpi| .{ dpi, "dpi" },
            .dpcm => |dpcm| .{ dpcm, "dpcm" },
            .dppx => |dppx| if (dest.targets.isCompatible(.x_resolution_unit))
                .{ dppx, "x" }
            else
                .{ dppx, "dppx" },
        };

        return try css.serializer.serializeDimension(value, unit, dest);
    }

    pub fn addF32(this: This, _: std.mem.Allocator, other: f32) Resolution {
        return switch (this) {
            .dpi => |dpi| .{ .dpi = dpi + other },
            .dpcm => |dpcm| .{ .dpcm = dpcm + other },
            .dppx => |dppx| .{ .dppx = dppx + other },
        };
    }
};

const bun = @import("bun");
const std = @import("std");

File: src/css/values/resolution.rs
use crate::css_parser as css;
use crate::css_parser::{CssResult as Result, Maybe, Parser, PrintErr, Printer, Token};
use crate::values::number::CSSNumber;
use bun_core::strings;

/// A CSS `<resolution>` value.
#[derive(
    Clone,
    Copy,
    Debug,
    PartialEq,
    crate::generics::CssEql,
    crate::generics::CssHash,
    crate::generics::DeepClone,
)]
pub enum Resolution {
    /// A resolution in dots per inch.
    Dpi(CSSNumber),
    /// A resolution in dots per centimeter.
    Dpcm(CSSNumber),
    /// A resolution in dots per px.
    Dppx(CSSNumber),
}

// ~toCssImpl

impl Resolution {
    // css.implementHash / css.implementEql — provided by
    // `#[derive(CssHash, CssEql)]` above (f32-payload enum).

    pub fn parse(input: &mut Parser) -> Result<Resolution> {
        // TODO: calc?
        let location = input.current_source_location();
        let tok = input.next()?.clone();
        if let Token::Dimension(dim) = &tok {
            let value = dim.num.value;
            let unit = dim.unit;
            return crate::match_ignore_ascii_case! { unit, {
                b"dpi" => Ok(Resolution::Dpi(value)),
                b"dpcm" => Ok(Resolution::Dpcm(value)),
                b"dppx" | b"x" => Ok(Resolution::Dppx(value)),
                _ => Err(location.new_unexpected_token_error(Token::Ident(unit))),
            }};
        }
        Err(location.new_unexpected_token_error(tok))
    }

    pub fn try_from_token(token: &Token) -> Maybe<Resolution, ()> {
        match token {
            Token::Dimension(dim) => {
                let value = dim.num.value;
                let unit = dim.unit;
                crate::match_ignore_ascii_case! { unit, {
                    b"dpi" => Ok(Resolution::Dpi(value)),
                    b"dpcm" => Ok(Resolution::Dpcm(value)),
                    b"dppx" | b"x" => Ok(Resolution::Dppx(value)),
                    _ => Err(()),
                }}
            }
            _ => Err(()),
        }
    }

    pub fn to_css(&self, dest: &mut Printer) -> core::result::Result<(), PrintErr> {
        let (value, unit): (CSSNumber, &'static [u8]) = match *self {
            Resolution::Dpi(dpi) => (dpi, b"dpi".as_slice()),
            Resolution::Dpcm(dpcm) => (dpcm, b"dpcm".as_slice()),
            Resolution::Dppx(dppx) => {
                if dest
                    .targets
                    .is_compatible(crate::compat::Feature::XResolutionUnit)
                {
                    (dppx, b"x".as_slice())
                } else {
                    (dppx, b"dppx".as_slice())
                }
            }
        };

        css::serializer::serialize_dimension(value, unit, dest)
    }

    pub fn add_f32(self, other: f32) -> Resolution {
        match self {
            Resolution::Dpi(dpi) => Resolution::Dpi(dpi + other),
            Resolution::Dpcm(dpcm) => Resolution::Dpcm(dpcm + other),
            Resolution::Dppx(dppx) => Resolution::Dppx(dppx + other),
        }
    }
}

// ported from: src/css/values/resolution.zig
